scripts/notebook_text_checker.py

```diff
@@ -0,0 +1,284 @@
+            )
+        ]
+
+    source = cell_source(first)
+
+    if TITLE_RE.fullmatch(source.strip()) is not None:
+        # all good, only whitespace and the # ... title.
+        return []
+
+    title = TITLE_RE.search(source)
+    if title is None:
+        return [
+            message_with_line(
+                first,
+                "The first cell be just the title for the notebook (like `# ...`) but it seems to be missing here",
+            )
+        ]
+
+    # report the error nicely
+    first_trailing_nonwhitespace = NON_WHITESPACE_RE.search(source, title.end())
+    if first_trailing_nonwhitespace is None:
+        end = len(source)
+    else:
+        end = first_trailing_nonwhitespace.end()
+
+    return [
+        message_with_line(
+            first,
+            "The first cell should contain only the title (like `# ...`) for the notebook. Additional introductory content can be in a separate following cell",
+            start=0,
+            end=end,
+        )
+    ]
+
+
+@checker
+def other_headings(notebook):
+    """
+    There should be no other H1/titles in the notebook, so that table of contents levels are correct
+    on Read the Docs.
+    """
+    previous_heading_level = 1
+
+    errors = []
+    for cell in notebook.cells[1:]:
+        if cell.cell_type != "markdown":
+            continue
+
+        source = cell_source(cell)
+        for heading in HEADING_RE.finditer(source):
+            if TITLE_RE.match(heading[0]):
+                errors.append(
+                    message_with_line(
+                        cell,
+                        "Found another title (like `# ...`) in internal cell. Later sections should use a high level heading (like `## ...` or `### ...`)",
+                        start=heading.start(),
+                    )
+                )
+
+            level = len(heading["level"])
+            highest_valid_level = previous_heading_level + 1
+            if level > highest_valid_level:
+                previous = "#" * previous_heading_level
+                suggestions = ", ".join(
+                    f"`{'#' * i} ...`" for i in range(2, highest_valid_level + 1)
+                )
+                errors.append(
+                    message_with_line(
+                        cell,
+                        f"Found a heading H{level} that skips level(s) from previous heading (H{previous_heading_level} `{previous} ...`). Consider using: {suggestions}",
+                        start=heading.start(),
+                    )
+                )
+
+            previous_heading_level = level
+
+    return errors
+
+
+# ANSI terminal escape sequences
+YELLOW_BOLD = "\033[1;33;40m"
+LIGHT_RED_BOLD = "\033[1;91;40m"
+RESET = "\033[0m"
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description="Validates that the description in notebooks follows the expected message_with_line, so that they read consistently and render nicely."
```

Comment:
Oops, search and replace gone off the rails! I rewrote this one a little bit more (but used `check` in the `locations` one):

```suggestion
        description="Validates that the descriptions in notebooks follow the expected format, so that the notebooks read consistently and render nicely."
```